src/pages/index.js

```diff
@@ -41,48 +51,110 @@ const enableValidation = (config) => {
 };
 enableValidation(config);
 
-const createCard = (cardObject) => {
-  const card = new Card(cardObject, "#card__template", () => {
-    imagePopup.open(cardObject);
+const handleImageClick = (data) => {
+  imagePopup.open(data);
+};
+
+const handleDelete = (card) => {
+  confirmationPopup.open();
+  confirmationPopup.setSubmitAction(() => {
+    api
+      .deleteCard(card)
+      .then(() => {})
+      .catch((err) => console.error(err));
+    confirmationPopup.close();
+    card.handleTrashBtn();
   });
+};
+
+const handleLike = (data, specificCard) => {
+  api
+    .likeCard(data)
+    .then((results) => {
+      specificCard.setLike(results.isLiked);
+    })
+    .catch((err) => console.error(err));
+};
+
+const createCard = (data) => {
+  const card = new Card(
+    data,
+    "#card__template",
+    handleImageClick,
+    handleDelete,
+    handleLike
+  );
   return card.generateCard();
 };
 
 const imagePopup = new PopupWithImage(".picture-modal");
 imagePopup.setEventListener();
 
-const cardSection = new Section(
-  {
-    data: initialCards,
-    renderer: (data) => {
-      const cardElement = createCard(data);
-      cardSection.addItem(cardElement);
-    },
-  },
-  cardsList
-);
-cardSection.renderItems();
-
 const newCardPopup = new PopupWithForm(".add-card-modal", (data) => {
-  const cardElement = createCard(data);
-  cardSection.addItem(cardElement);
+  newCardPopup.renderLoading(true);
   newCardPopup.close();
+  api
+    .addNewCard(data)
+    .then((results) => {
+      const cardElement = createCard(results);
+      cardSection.addItem(cardElement);
+    })
+    .catch((err) => console.error(err))
+    .finally(() => {
+      newCardPopup.renderLoading(false);
+    });
 });
 newCardPopup.setEventListener();
 
-const userInfo = new UserInfo(".profile__name", ".profile__occupation");
+const userInfo = new UserInfo(
+  ".profile__name",
+  ".profile__occupation",
+  ".profile__avatar"
+);
 
 const profilePopup = new PopupWithForm(".profile-modal", (data) => {
+  profilePopup.renderLoading(true);
```

Comment:
**Could be improved**

If it’s interesting for you here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **closing** and **catching errors**


```js

// You can make a universal function that accepts a request, popup instance and optional loading text  
function handleSubmit(request, popupInstance, loadingText = "Saving...") {

  // here we change the button text
  popupInstance.renderLoading(true, loadingText);
  request()
    .then(() => {
      // We need to close only in `then`
      popupInstance.close()
    })
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
     .catch(console.error)
    // in `finally` we need to return the initial button text back in any case
    .finally(() => {
      popupInstance.renderLoading(false);
    });
}



```


You need to pass the **request** into the call of `handleSubmit` like that:



```js

// here is an example of the profile form handling
function handleProfileFormSubmit(inputValues) {
  // we create a function that returns a promise
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally` inside `handleSubmit`
    return api.editProfile(inputValues).then((userData) => {
      userInfo.setUserInfo(userData)
    });
  }
  // Here we call the function passing the request, popup instance and if we need some other loading text we can pass it as the 3rd argument
  handleSubmit(makeRequest, profilePopup);
}



```